### PR TITLE
Version symbols to maintain ABI compatibility

### LINKED ISSUE #115: SONAME bump?
Since commit 50b3ba132d22 changed the signatures of many public libkcapi functions, shouldn't this be accompanied by an SONAME bump of the library? Looking at the build scripts, it seems this would be achieved by bumping the major version.

I'm not aware of any packages at least in Fedora that would link against libkcapi, but still it would be better to follow the correct procedure.

### PR DESCRIPTION
**Pull request by simo5**
Before 1.3.0 many functions had a different signature.
Add back compatibility functions and actually use symbol versioning to
maintain binary compatibility.

Note that the 1.3.0 release is just broken any build against that
version will break any previous or later built binaries that link to it,
and there is nothing that can be done to fix that version.

Fixes #115
Although it does that not via soname bump but by actually using symbol versioning for what it was intended :)

### CODE REVIEW

File: lib/internal.h [hunk 1]
```diff
@@ -349,10 +349,24 @@ static inline int io_getevents(__attribute__((unused)) aio_context_t ctx,
...
+    __attribute__ ((visibility ("default")))
+
+# define ORIG_SYMVER(name, version) \
+    __asm__(".global orig_" #name ";"\
+	    ".symver orig_" #name ",kcapi_" #name "@LIBKCAPI_" version);\
+    __attribute__ ((visibility ("default")))
+
 #else
 # define DSO_PUBLIC
+# define IMPL_SYMVER(name, version) typedef impl_##name kcapi_##name;
```
Comment (WOnder93) [inline lib/internal.h:365]: Doesn't `typedef` work only on types?
https://gcc.godbolt.org/z/fMhc8646W

Comment (simo5) [reply lib/internal.h:365]: Eh that part I hastly put together, not sure it is correct indeed

File: lib/version.lds [hunk 2]
```diff
@@ -160,3 +160,69 @@ LIBKCAPI_1.3.0 {
...
+LIBKCAPI_1.3.1 {
+	kcapi_cipher_encrypt;
+	kcapi_cipher_encrypt_aio;
+	kcapi_cipher_decrypt;
+	kcapi_cipher_decrypt_aio;
+	kcapi_cipher_stream_init_enc;
+	kcapi_cipher_stream_init_dec;
+	kcapi_cipher_stream_update;
+	kcapi_cipher_stream_op;
+	kcapi_aead_getdata_input;
```
Comment (smuellerDD) [inline lib/version.lds:173]: please add kcapi_cipher_[enc|dec]_aes_[cbc|ctr]

File: lib/version.lds [hunk 3]
```diff
@@ -160,3 +160,69 @@ LIBKCAPI_1.3.0 {
...
+	kcapi_aead_getdata_output;
```
Comment (smuellerDD) [inline lib/version.lds:174]: please add kcapi_aead_setassoclen

File: lib/version.lds (hunk 2 above)
Comment (simo5) [reply lib/version.lds:173]: oops, thanks for catching it

**Comment by simo5** [comment]
fixed missing symbols as well as fixing the actual version on the aes ones, it was set to 0.12.0 and correct to 1.0.0 now

**Comment by smuellerDD** [comment]
The rest looks good - not sure what we do about IMPL_SYMVER though.

**Comment by simo5** [comment]
> The rest looks good - not sure what we do about IMPL_SYMVER though.

Working on that, give me a little time

**Comment by simo5** [comment]
@smuellerDD do we really need to support gcc < 4 anymore?
If not we can simply remove the fallback and be done with it

**Comment by smuellerDD** [comment]
Let's remove it - gcc 4 - who uses that :-)
Could you just add an #error instead?

File: lib/internal.h (hunk 1 above)
Comment (simo5) [reply lib/internal.h]: removed

**Comment by simo5** [comment]
Added #error for gcc < 4
Does that work for you ?

**Comment by smuellerDD** [comment]
Looks good - once you pushed the update on rng_get_bytes, I will merge it

Thanks!

**Comment by simo5** [comment]
Updated version for rng_get_bytes and cipher_stream_update_last

**Comment by smuellerDD** [comment]
Build error:

```
libtool: compile:  gcc -DPACKAGE_NAME=\"libkcapi\" -DPACKAGE_TARNAME=\"libkcapi\" -DPACKAGE_VERSION=\"1.3.1\" "-DPACKAGE_STRING=\"libkcapi 1.3.1\"" -DPACKAGE_BUGREPORT=\"\" -DPACKAGE_URL=\"\" -DKCAPI_MAJVERSION=1 -DKCAPI_MINVERSION=3 -DKCAPI_PATCHLEVEL=1 -DPACKAGE=\"libkcapi\" -DVERSION=\"1.3.1\" -DSTDC_HEADERS=1 -DHAVE_SYS_TYPES_H=1 -DHAVE_SYS_STAT_H=1 -DHAVE_STDLIB_H=1 -DHAVE_STRING_H=1 -DHAVE_MEMORY_H=1 -DHAVE_STRINGS_H=1 -DHAVE_INTTYPES_H=1 -DHAVE_STDINT_H=1 -DHAVE_UNISTD_H=1 -DHAVE_DLFCN_H=1 -DLT_OBJDIR=\".libs/\" -DHAVE_GETRANDOM=1 -DHAVE_SECURE_GETENV=1 -DWITH_LIB_ASYM=1 -DWITH_LIB_KPP=1 -DCHECK_PREFIX=\".\" -DCHECK_SUFFIX=\"hmac\" -DHAVE_FSEEKO=1 -I. -Wextra -Wall -pedantic -fwrapv --param ssp-buffer-size=4 -O2 -Werror -std=gnu99 -Wconversion -fvisibility=hidden -D_FORTIFY_SOURCE=2 -fstack-protector-strong -g -O2 -DPIE -MT lib/libkcapi_la-kcapi-asym.lo -MD -MP -MF lib/.deps/libkcapi_la-kcapi-asym.Tpo -c lib/kcapi-asym.c  -fPIC -DPIC -o lib/.libs/libkcapi_la-kcapi-asym.o
/tmp/ccODkSNy.s: Assembler messages:
/tmp/ccODkSNy.s: Error: invalid attempt to declare external version name as default in symbol `kcapi_akciper_encrypt_aio@@LIBKCAPI_1.3.1'
```

**Comment by smuellerDD** [comment]
It is a typo, I have fixed it

**Comment by simo5** [comment]
the name is spelled correctly in my tree afaik

**Comment by smuellerDD** [comment]
All tests pass. Closing.

Thanks a whole lot for your support!

**Pull request closed without merging**